File: cpp/utils/timer.cpp
#include <iostream>
#include <stdexcept>
#include "timer.h"

Timer::Timer()
: m_start(std::chrono::high_resolution_clock::now())
, m_reported(false)
{
}

Timer::~Timer() {
    std::chrono::high_resolution_clock::time_point end = std::chrono::high_resolution_clock::now();
    std::chrono::nanoseconds elapsed = end - m_start;
    std::cerr
        << "Time elapsed: "
        << std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count()
        << "µs"
        << std::endl;
    m_reported = true;
}

File: rs/utils/src/timer.rs
use std::time::Instant;

pub struct Timer {
    start: std::time::Instant,
}

impl Timer {
    pub fn new() -> Self {
        Self {
            start: Instant::now()
        }
    }
}

impl Default for Timer {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for Timer {
    fn drop(&mut self) {
        eprintln!("Elapsed: {}µs", self.start.elapsed().as_micros());
    }
}
